Wangscape/image/Image.cpp

```diff
@@ -30,3 +63,254 @@ ImageGrey imageToGreyscale(const ImageColour & image)
         logWarning() << "Image is not greyscale; using red channel";
     return image_r;
 }
+
+ImageColour imageGreyscaleToImage(const ImageGrey & image)
+{
+    ImageColour image_colour(image.n_rows, image.n_cols, 4);
+    image_colour.slice(0) = image;
+    image_colour.slice(1) = image;
+    image_colour.slice(2) = image;
+    image_colour.slice(3).fill(255);
+    return image_colour;
+}
+
+sf::Image imageToSFImage(const ImageColour & image)
+{
+    sf::Image sf_image;
+    sf_image.create(image.n_cols, image.n_rows);
+    UVec shape = makeUVec(sf_image.getSize());
+    for (size_t y = 0; y < shape.y(); y++)
+        for (size_t x = 0; x < shape.x(); x++)
+        {
+            const auto pixel = image.tube(y, x);
+            sf_image.setPixel(x, y, sf::Color(pixel(0), pixel(1), pixel(2), pixel(3)));
+        }
+    return sf_image;
+}
+
+void copyRegionBounded(const ImageGrey& source, ImageGrey & target, IVec source_origin, IVec target_origin, IVec size)
+{
+    const auto source_shape = makeUVec(arma::size(source));
+    const auto target_shape = makeUVec(arma::size(target));
+    // Cut off negative indices in source or target
+    IVec origin_excess;
+    origin_excess.x() = std::min({source_origin.x(),
+                                target_origin.x(),
+                                0});
+    origin_excess.y() = std::min({source_origin.y(),
+                                target_origin.y(),
+                                0});
+    source_origin -= origin_excess;
+    target_origin -= origin_excess;
+    size += origin_excess;
+    // Cut off indices greater than shape in source or target
+    IVec end_excess;
+    const auto source_end = source_origin + size;
+    const auto target_end = target_origin + size;
+    end_excess.x() = std::max({source_end.x() - (int)source_shape.x(),
+                             target_end.x() - (int)target_shape.x(),
+                             0});
+    end_excess.y() = std::max({source_end.y() - (int)source_shape.y(),
+                             target_end.y() - (int)target_shape.y(),
+                             0});
+    size -= end_excess;
```

Comment:
I don't think there's much reason to use these two parts of the function separately. However, everything except the `CopyRegion` call should be moved into a separate function to make testing easier - at the moment this can only be tested by checking the output array.